### PR TITLE
fix: support `date - timestamp` and `timestamp - date`

### LINKED ISSUE #5931: Data Timestamp don't support minus
### Describe the bug

Data Timestamp don't support minus

### To Reproduce

SELECT DATE '2023-04-09' - DATE '2023-04-02';

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #5931 .

# Rationale for this change

support data/timestamp minus in type coercion.

# What changes are included in this PR?

Add test in slt.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/dates.slt [hunk 1]
```diff
@@ -88,3 +88,19 @@ h
 statement error DataFusion error: Error during planning: Timestamp\(Nanosecond, Some\("\+00:00"\)\) \+ Utf8 can't be evaluated because there isn't a common type to coerce the types to
 select i_item_desc from test
 where d3_date > now() + '5 days';
+
+# DATE minus DATE
```
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/core/tests/sqllogictests/test_files/dates.slt:92]: Great, btw we already have `array_typeof.slt` which also covers some casts

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 2]
```diff
@@ -660,6 +660,12 @@ statement error DataFusion error: Error during planning: Timestamp\(Nanosecond,
 SELECT ts1 + ts2
 FROM foo;
 
+# Timestamp - Timestamp
+query ?
```
Comment (comphead) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:664]: Please add negative test if older date - newer date

File: datafusion/expr/src/type_coercion/binary.rs [hunk 3]
```diff
@@ -239,8 +238,7 @@ pub fn temporal_add_sub_coercion(
             if (is_date(lhs) || is_timestamp(lhs))
                 && (is_date(rhs) || is_timestamp(rhs)) =>
         {
-            // TODO: support Minus
-            None
+            temporal_coercion(lhs, rhs)
```
Comment (comphead) [inline datafusion/expr/src/type_coercion/binary.rs:241]: if the comment about error still relevant? 
for `temporal_coercion` if it does the actual substract?

**Review by ozankabak — APPROVED** [review]
Thank you @jackwener, LGTM. @berkaysynnada can chime in if there is anything I'm missing

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 4]
```diff
@@ -660,6 +660,19 @@ statement error DataFusion error: Error during planning: Timestamp\(Nanosecond,
...
+SELECT '2000-01-01T00:00:00'::timestamp - '2000-01-01T00:00:00'::timestamp;
+----
+0 years 0 mons 0 days 0 hours 0 mins 0.000000000 secs
+
+# large timestamp - small timestamp
+# Todo: Does it need to be fixed for negative values?.
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:670]: I think negative values is correct and a valid interval. The intervals are signed according to the arrow spec:

https://github.com/apache/arrow/blob/776ea3f4f3d665f38fc370d0f284760eafa7de45/format/Schema.fbs#L363-L384

File: datafusion/expr/src/type_coercion/binary.rs [hunk 5]
```diff
@@ -754,51 +751,6 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
     }
 }
 
-/// Coercion rule for numerical types: The type that both lhs and rhs
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:757]: it is always nice to see a bug fixed by deleting code. Thank you @jackwener  👏

**Review by alamb — APPROVED** [review]
Thanks @jackwener  and @comphead

**Pull request merged**